File: src/dnfw/mods/arpmodes.py
"""Two more arpeggiator modes: SHUF and RAND, whose names the firmware already has.

Stock Digitone II 1.11 names eight arp modes, `OFF TRUE UP DOWN CYCL SHUF RAND
CHRD`, offers five, and plays CYCL for anything above DOWN
(`docs/arp-hidden-modes.md`). This mod gives SHUF and RAND real code and
widens the MODE menu to them; CHRD (7) stays out, and a 7 still plays CYCL.

- **SHUF**: every note of the range (the held notes x RNG + 1 octaves) once per
  cycle, in a new random order each cycle; the note that ends one cycle never
  starts the next.
- **RAND**: a random note of the range on every step, repeats allowed.

SPD, LEN, the step mask (a muted step rests and does not advance), the step
offsets and N.LEN behave as in every other mode: the new code only chooses the
note and octave and hands them to the step's own tail.

`scripts/build_arpmodes.py` has the design, the RAM layout and the choice of
random source (a private xorshift32, stirred with the millisecond tick when an
arp starts; the firmware's shared `rand()` is not touched).

## How it applies

The code is assembled ahead of time (`scripts/gen_arpmodes_code.py` ->
`arpmodes_code.json`, which is checked to reproduce the build byte for byte):

1. every guard and every edit's stock bytes are checked -- an image that
   differs is refused;
2. the edits are written: two code caves, the dispatch hook, five bounds.

It writes only inside section 3, changes no length and appends nothing. RAM:
2,320 bytes at `0x467a0000`, above BSS.

## A sound saved with SHUF or RAND

Saved as MODE 5 or 6. On stock firmware (or any build without this mod) it
loads with its arp **OFF**: the stock LOAD bound turns anything above 4 into 0.

## With arpplocks

arpplocks' MODE lock takes its ceiling from the menu edit's own clamp (the
immediate of `0x4004bf00`), so with this mod a MODE lock reaches RAND, and
without it the lock stops at CYCL as before. Nothing here writes arpplocks'
bytes, and the two apply in either order.
"""

from __future__ import annotations

import json
import pathlib

from . import Extent, ModError, Result

ID = "arpmodes"
NAME = "Arpeggiator SHUF and RAND"
SUMMARY = "Two more arp modes: SHUF (each note once per cycle, shuffled) and RAND."
DEVICE = 0x15                      # Digitone II
SECTION = 3                        # MAIN OS
BASE = 0x40000400
SPEC = json.loads((pathlib.Path(__file__).with_name("arpmodes_code.json")).read_text())


def extents(firmware=None) -> list[Extent]:
    return [Extent(SECTION, e["va"] - BASE, len(e["new"]) // 2, e["what"]) for e in SPEC["edits"]]


def apply(firmware) -> Result:
    section = firmware.container.find(SECTION)
    if section is None:
        raise ModError("image has no MAIN OS section")
    original = section.unpack()
    if original is None:
        raise ModError("MAIN OS did not depack")
    # Longer is fine: data appended after the stock end (lfowaves, bootscreen)
    # moves no address this mod writes or reads. The guards identify the build.
    if len(original) < SPEC["stock_length"]:
        raise ModError(f"MAIN OS is {len(original):,} B, shorter than "
                       f"{SPEC['stock_length']:,}: not Digitone II 1.11")
    for g in SPEC["guards"]:
        want = bytes.fromhex(g["bytes"])
        at = g["va"] - BASE
        if original[at:at + len(want)] != want:
            raise ModError(f"0x{g['va']:08x} ({g['what']}) is not stock; this mod is "
                           "for unmodified Digitone II 1.11")
    for e in SPEC["edits"]:
        want = bytes.fromhex(e["stock"])
        at = e["va"] - BASE
        if original[at:at + len(want)] != want:
            raise ModError(f"0x{e['va']:08x} ({e['what']}) is not stock; this mod is "
                           "for unmodified Digitone II 1.11, or another mod already "
                           "wrote there")

    content = bytearray(original)
    for e in SPEC["edits"]:
        new = bytes.fromhex(e["new"])
        content[e["va"] - BASE:e["va"] - BASE + len(new)] = new

    return Result(payloads={SECTION: bytes(content)}, extents=extents(),
                  notes=["the ARPEGGIATOR MODE menu offers SHUF and RAND after CYCL, "
                         "and stops there",
                         f"{len(SPEC['edits'])} edits in section 3, nothing appended"])

File: site/js/mods/arpmodes.js
/**
 * Two more arpeggiator modes: SHUF and RAND, whose names the firmware already has.
 *
 * The browser half of `src/dnfw/mods/arpmodes.py`, which carries the evidence and
 * the hardware result. Same steps, same order, same refusals: check the stock
 * length, every guard and every edit's stock bytes, then write the edits (two
 * code caves, the dispatch hook, five MODE bounds). Nothing is appended.
 * `scripts/js_arpmodes_check.mjs` fails if the bytes differ from the Python's,
 * and the Python's are checked byte for byte against the build when
 * `scripts/gen_arpmodes_code.py` packages it.
 */

import { CODE } from "./arpmodes-code.js";

export const ID = "arpmodes";
export const NAME = "Arpeggiator SHUF and RAND";
export const SECTION = 3;
const BASE = 0x40000400;

export class ModError extends Error {}

const hex = (s) => Uint8Array.from(s.match(/../g) ?? [], (b) => parseInt(b, 16));
const toHex = (a) => Array.from(a, (b) => b.toString(16).padStart(2, "0")).join("");
/** Python's `f"{n:,}"`, whatever the browser's locale. */
const grouped = (n) => String(n).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
/** Python's `f"0x{va:08x}"`. */
const va8 = (va) => `0x${va.toString(16).padStart(8, "0")}`;

export const EDITS = CODE.edits.length;
export const EDITED_BYTES = CODE.edits.reduce((n, e) => n + e.new.length / 2, 0);
export const CAVES = CODE.edits.filter((e) => e.what === "arp modes code cave");
export const RAM_BYTES = CODE.ram.reduce((n, r) => n + r.bytes, 0);

export function extents() {
  return CODE.edits.map((e) => ({ section: SECTION, start: e.va - BASE,
                                  length: e.new.length / 2, what: e.what }));
}

/** Section 3 of `firmware`, refused unless this mod applies to it. */
export function check(firmware) {
  const section = firmware.container.find(SECTION);
  if (section === null || section === undefined) throw new ModError("image has no MAIN OS section");
  const original = section.unpack();
  if (original === null || original === undefined) throw new ModError("MAIN OS did not depack");
  // Longer is fine: data appended after the stock end (lfowaves, bootscreen)
  // moves no address this mod writes or reads. The guards identify the build.
  if (original.length < CODE.stock_length) {
    throw new ModError(`MAIN OS is ${grouped(original.length)} B, shorter than `
      + `${grouped(CODE.stock_length)}: not Digitone II 1.11`);
  }
  const at = (va, n) => toHex(original.subarray(va - BASE, va - BASE + n));
  for (const g of CODE.guards) {
    if (at(g.va, g.bytes.length / 2) !== g.bytes) {
      throw new ModError(`${va8(g.va)} (${g.what}) is not stock; this mod is `
        + "for unmodified Digitone II 1.11");
    }
  }
  for (const e of CODE.edits) {
    if (at(e.va, e.stock.length / 2) !== e.stock) {
      throw new ModError(`${va8(e.va)} (${e.what}) is not stock; this mod is `
        + "for unmodified Digitone II 1.11, or another mod already wrote there");
    }
  }
  return original;
}

/** -> `{ content, notes }`, the new section 3. */
export function apply(firmware) {
  const original = check(firmware);
  const content = original.slice();
  for (const e of CODE.edits) content.set(hex(e.new), e.va - BASE);
  return {
    content,
    notes: ["the ARPEGGIATOR MODE menu offers SHUF and RAND after CYCL, and stops there",
            `${CODE.edits.length} edits in section 3, nothing appended`],
  };
}
